engines/sci/sound/drivers/macmixer.h

```diff
@@ -0,0 +1,268 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "audio/audiostream.h"
+#include "audio/mixer.h"
+#include "common/frac.h"
+#include "common/system.h"
+
+#ifndef SCI_SOUND_DRIVERS_MACMIXER_H
+#define SCI_SOUND_DRIVERS_MACMIXER_H
+
+namespace Sci {
+
+// Unsigned version of frac_t
+typedef uint32 ufrac_t;
+static inline ufrac_t uintToUfrac(uint16 value) { return value << FRAC_BITS; }
+static inline uint16 ufracToUint(ufrac_t value) { return value >> FRAC_BITS; }
+
+template <typename T>
+class Mixer_Mac : public Audio::AudioStream {
+public:
+	enum {
+		kChannels = 4,
+		kInterruptFreq = 60
+	};
+
+	enum Mode {
+		kModeAuthentic,
+		kModeHq,
+		kModeHqStereo
+	};
+
+	Mixer_Mac(Mode mode);
+	void startMixer();
+	void stopMixer();
+	void setMixerVolume(byte volume) { _mixVolume = volume; }
+	void resetChannel(uint channel);
+	void resetChannels();
+	// NOTE: Last sample accessed is data[endOffset + 1] in kModeHq(Stereo)
+	void setChannelData(uint channel, const byte *data, uint16 startOffset, uint16 endOffset, uint16 loopLength = 0);
+	void setChannelStep(uint channel, ufrac_t step);
+	void setChannelVolume(uint channel, byte volume);
+	void setChannelPan(uint channel, byte pan);
+
+	// AudioStream
+	bool isStereo() const override { return _mode == kModeHqStereo; }
+	int getRate() const override { return (_mode == kModeAuthentic ? 11127 : g_system->getMixer()->getOutputRate()); }
+	int readBuffer(int16 *data, const int numSamples) override;
+	bool endOfData() const override { return false; }
+
+private:
+	template <Mode mode>
+	void generateSamples(int16 *buf, int len);
+
+	struct Channel {
+		ufrac_t pos;
+		ufrac_t step;
+		const byte *data;
+		uint16 endOffset;
+		uint16 loopLength;
+		byte volume;
+		int8 pan;
+	};
+
+	ufrac_t _nextTick;
+	ufrac_t _samplesPerTick;
+	bool _isPlaying;
+	const Mode _mode;
+	Channel _mixChannels[kChannels];
+	byte _mixVolume;
+};
+
+template <typename T>
+Mixer_Mac<T>::Mixer_Mac(Mixer_Mac::Mode mode) :
```

Comment:
VS2015 can't compile this line, it gets confused by the "Mixer_Mac::" qualifier on the parameter because it's an enum in the template class. Removing "Mixer_Mac::" avoids this while keeping the correct scope on the Mode type.

I can't find a VS bug report for this, but I verified it in a standalone test case and that it was fixed in at least VS2019.
Sorry I use an old one! =)

error C2988: unrecognizable template declaration/definition